File: libs/power-log-parser/src/lib/parsers/rumble-run-step-parser.ts
import { GameTag, ScenarioId } from '@firestone-hs/reference-data';
import { ActionParser } from '../action-parser';
import { GameEventProvider } from '../game-event';
import { Action, Node, NodeType } from '../models';
import { GameState } from '../state/game-state';
import { ParserState, StateType } from '../state/parser-state';
import type { StateFacade } from '../state/state-facade';
import { actionQualifiesForRumbleRunStepAction, lastHeroHealthTagChangeInAction } from './pve-run-step-health-passive';

const STARTING_HEALTH = 20;

export class RumbleRunStepParser implements ActionParser {
	readonly ParserName = 'RumbleRunStepParser';

	private GameState: GameState;
	private ParserState: ParserState;
	private StateFacade: StateFacade;

	constructor(parserState: ParserState, stateFacade: StateFacade) {
		this.ParserState = parserState;
		this.GameState = parserState.GameState;
		this.StateFacade = stateFacade;
	}

	AppliesOnNewNode(_node: Node, _stateType: StateType): boolean {
		return false;
	}

	AppliesOnCloseNode(node: Node, stateType: StateType): boolean {
		if (stateType !== StateType.PowerTaskList || node.Type !== NodeType.Action) {
			return false;
		}
		const action = node.Object as Action;
		const heroEntityId = this.resolveLocalHeroEntityId();
		if (heroEntityId == null) {
			return false;
		}
		return actionQualifiesForRumbleRunStepAction(action, heroEntityId, this.ParserState.GameState.GameEntityId);
	}

	CreateGameEventProviderFromNew(_node: Node): GameEventProvider[] | null {
		return null;
	}

	CreateGameEventProviderFromClose(node: Node): GameEventProvider[] | null {
		return [
			GameEventProvider.Create(
				(node.Object as Action).TimeStamp,
				'RUMBLE_RUN_STEP',
				() => {
					if (this.StateFacade.ScenarioID !== (ScenarioId.TRLA_DUNGEON as number)) {
						return null;
					}
					const action = node.Object as Action;
					const heroEntityId = this.resolveLocalHeroEntityId();
					if (heroEntityId == null) {
						return null;
					}
					if (
						!actionQualifiesForRumbleRunStepAction(
							action,
							heroEntityId,
							this.ParserState.GameState.GameEntityId,
						)
					) {
						return null;
					}
					const tagChange = lastHeroHealthTagChangeInAction(action, heroEntityId);
					const healthChangeDef =
						(tagChange != null
							? tagChange.Value
							: this.ParserState.GetEntity(heroEntityId)!.GetTag(GameTag.HEALTH)) - STARTING_HEALTH;
					const runStep = 1 + Math.floor(healthChangeDef / 5);
					return {
						Type: 'RUMBLE_RUN_STEP',
						Value: runStep,
					};
				},
				true,
				node,
			),
		];
	}

	private resolveLocalHeroEntityId(): number | null {
		const localPlayerId = this.StateFacade.LocalPlayer?.Id;
		if (localPlayerId == null) {
			return null;
		}
		const playerEntity = this.ParserState.GetEntity(localPlayerId);
		if (playerEntity == null) {
			return null;
		}
		const heroEntityId = playerEntity.GetTag(GameTag.HERO_ENTITY);
		return heroEntityId > 0 ? heroEntityId : null;
	}
}

File: tools/linux-helpers/hs-converter/HearthstoneReplays/Events/Parsers/RumbleRunStepParser.cs
using HearthstoneReplays.Parser;
using HearthstoneReplays.Parser.ReplayData;
using HearthstoneReplays.Parser.ReplayData.GameActions;
using System;
using HearthstoneReplays.Enums;
using HearthstoneReplays.Parser.ReplayData.Entities;
using System.Collections.Generic;
using System.Linq;

namespace HearthstoneReplays.Events.Parsers
{
    public class RumbleRunStepParser : ActionParser
    {
        private static readonly int STARTING_HEALTH = 20;

        private GameState GameState { get; set; }
        private ParserState ParserState { get; set; }
        private StateFacade Helper { get; set; }

        public RumbleRunStepParser(ParserState ParserState, StateFacade helper)
        {
            this.ParserState = ParserState;
            this.GameState = ParserState.GameState;
            this.Helper = helper;
        }

        public bool AppliesOnNewNode(Node node, StateType stateType)
        {
            return false;
            //return node.Type == typeof(TagChange)
            //    && (node.Object as TagChange).Name == (int)GameTag.HEALTH
            //    && !string.IsNullOrWhiteSpace((node.Object as TagChange).DefChange);
        }

        public bool AppliesOnCloseNode(Node node, StateType stateType)
        {
            return stateType == StateType.PowerTaskList
                && node.Type == typeof(Parser.ReplayData.GameActions.Action)
                && (node.Object as Parser.ReplayData.GameActions.Action).Data
                    .Where(data => data.GetType() == typeof(TagChange))
                    .Select(data => data as TagChange)
                    .Where(tag => tag.Name == (int)GameTag.HEALTH && !string.IsNullOrWhiteSpace(tag.DefChange))
                    .Count() > 0;
        }

        public List<GameEventProvider> CreateGameEventProviderFromNew(Node node)
        {
            return null;
        }

        public List<GameEventProvider> CreateGameEventProviderFromClose(Node node)
        {
            return new List<GameEventProvider> { GameEventProvider.Create(
                (node.Object as Parser.ReplayData.GameActions.Action).TimeStamp,
                "RUMBLE_RUN_STEP",
                () => {
                    if (Helper.ScenarioID != (int)Scenario.RUMBLE_RUN)
                    {
                        return null;
                    }
                    var action = (node.Object as Parser.ReplayData.GameActions.Action);
                    var heroEntityId = ParserState.GetEntity(Helper.LocalPlayer.Id).GetTag(GameTag.HERO_ENTITY);
                    var tagChange = action.Data
                            .Where(data => data.GetType() == typeof(TagChange))
                            .Select(data => data as TagChange)
                            .Where(tag => tag.Name == (int)GameTag.HEALTH && !string.IsNullOrWhiteSpace(tag.DefChange))
                            .Where(tag => tag.Entity == heroEntityId)
                            .FirstOrDefault();
                    // If no tag change for the main player, we use the starting health
                    // The issue with computing things this way is that it's now possible
                    // to emit the event multiple times over the course of a match
                    var healthChangeDef = (tagChange != null ? tagChange.Value : ParserState.GetEntity(heroEntityId).GetTag(GameTag.HEALTH)) - STARTING_HEALTH;
                    var runStep = 1 + healthChangeDef / 5;
                    return new GameEvent
                    {
                        Type = "RUMBLE_RUN_STEP",
                        Value = runStep
                    };
                },
                true,
                node) };
        }
    }
}
